File: src/Application Framework/Entities/Management/SettingsValidator.cs
using System;
using System.Collections;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace FSM.Entity
{
    namespace Management
    {
        public class SettingsValidator : BaseValidator
        {
            public void Validate(Settings settings)
            {
                if (settings.Errors.Count > 0)
                {
                    DictionaryEntry de;
                    foreach (var de in settings.Errors)
                        AddCriticalMessage(Conversions.ToString(de.Value));
                }

                bool checkDay = true;
                if (settings.WorkingHoursStart.Trim().Length == 0)
                {
                    AddCriticalMessage("Working Hours Start Missing");
                    checkDay = true;
                }

                if (settings.WorkingHoursEnd.Trim().Length == 0)
                {
                    AddCriticalMessage("Working Hours End Missing");
                    checkDay = true;
                }

                if (checkDay)
                {
                    var sd = new DateTime(DateAndTime.Now.Year, DateAndTime.Now.Month, DateAndTime.Now.Day, Conversions.ToInteger(settings.WorkingHoursStart.Trim().Substring(0, 2)), Conversions.ToInteger(settings.WorkingHoursStart.Trim().Substring(3, 2)), 0);
                    var ed = new DateTime(DateAndTime.Now.Year, DateAndTime.Now.Month, DateAndTime.Now.Day, Conversions.ToInteger(settings.WorkingHoursEnd.Trim().Substring(0, 2)), Conversions.ToInteger(settings.WorkingHoursEnd.Trim().Substring(3, 2)), 0);
                    if (sd >= ed)
                    {
                        AddCriticalMessage("Working Hours End must be after the Start");
                    }
                }

                if (settings.MileageRate == 0)
                {
                    AddCriticalMessage("Mileage Rate Missing");
                }

                if (!Information.IsNumeric(settings.MileageRate))
                {
                    AddCriticalMessage("Mileage Rate Invalid");
                }

                if (settings.PartsMarkup == 0)
                {
                    AddCriticalMessage("Parts Markup Missing");
                }

                if (!Information.IsNumeric(settings.PartsMarkup))
                {
                    AddCriticalMessage("Parts Markup Invalid");
                }

                if (settings.RatesMarkup == 0)
                {
                    AddCriticalMessage("Rates Markup Missing");
                }

                if (!Information.IsNumeric(settings.RatesMarkup))
                {
                    AddCriticalMessage("Rates Markup Invalid");
                }

                if (settings.CalloutPrefix.Trim().Length == 0)
                {
                    AddCriticalMessage("Callout Prefix Missing");
                }

                if (settings.MiscPrefix.Trim().Length == 0)
                {
                    AddCriticalMessage("Miscellaneous Prefix Missing");
                }

                if (settings.PPMPrefix.Trim().Length == 0)
                {
                    AddCriticalMessage("PPM Prefix Missing");
                }

                if (settings.QuotePrefix.Trim().Length == 0)
                {
                    AddCriticalMessage("Quote Prefix Missing");
                }

                if (settings.ServiceFromLetterPrefix.Trim().Length == 0)
                {
                    AddCriticalMessage("Service From Letter Prefix Missing");
                }

                if (ValidatorMessages.CriticalMessages.Count > 0)
                {
                    throw new ValidationException(this);
                }
            }
        }
    }
}

File: src/Application Framework/Entities/Management/SettingsValidator.vb
Namespace Entity

    Namespace Management

        Public Class SettingsValidator : Inherits BaseValidator

            Public Sub Validate(ByVal settings As Entity.Management.Settings)
                If settings.Errors.Count > 0 Then
                    Dim de As DictionaryEntry
                    For Each de In settings.Errors
                        Me.AddCriticalMessage(CStr(de.Value))
                    Next
                End If

                Dim checkDay As Boolean = True
                If settings.WorkingHoursStart.Trim.Length = 0 Then
                    Me.AddCriticalMessage("Working Hours Start Missing")
                    checkDay = True
                End If
                If settings.WorkingHoursEnd.Trim.Length = 0 Then
                    Me.AddCriticalMessage("Working Hours End Missing")
                    checkDay = True
                End If
                If checkDay Then
                    Dim sd As DateTime = New DateTime(Now.Year, Now.Month, Now.Day, settings.WorkingHoursStart.Trim.Substring(0, 2), settings.WorkingHoursStart.Trim.Substring(3, 2), 0)
                    Dim ed As DateTime = New DateTime(Now.Year, Now.Month, Now.Day, settings.WorkingHoursEnd.Trim.Substring(0, 2), settings.WorkingHoursEnd.Trim.Substring(3, 2), 0)

                    If sd >= ed Then
                        Me.AddCriticalMessage("Working Hours End must be after the Start")
                    End If
                End If

                If settings.MileageRate = 0 Then
                    Me.AddCriticalMessage("Mileage Rate Missing")
                End If
                If Not IsNumeric(settings.MileageRate) Then
                    Me.AddCriticalMessage("Mileage Rate Invalid")
                End If
                If settings.PartsMarkup = 0 Then
                    Me.AddCriticalMessage("Parts Markup Missing")
                End If
                If Not IsNumeric(settings.PartsMarkup) Then
                    Me.AddCriticalMessage("Parts Markup Invalid")
                End If
                If settings.RatesMarkup = 0 Then
                    Me.AddCriticalMessage("Rates Markup Missing")
                End If
                If Not IsNumeric(settings.RatesMarkup) Then
                    Me.AddCriticalMessage("Rates Markup Invalid")
                End If

                If settings.CalloutPrefix.Trim.Length = 0 Then
                    Me.AddCriticalMessage("Callout Prefix Missing")
                End If
                If settings.MiscPrefix.Trim.Length = 0 Then
                    Me.AddCriticalMessage("Miscellaneous Prefix Missing")
                End If
                If settings.PPMPrefix.Trim.Length = 0 Then
                    Me.AddCriticalMessage("PPM Prefix Missing")
                End If
                If settings.QuotePrefix.Trim.Length = 0 Then
                    Me.AddCriticalMessage("Quote Prefix Missing")
                End If

                If settings.ServiceFromLetterPrefix.Trim.Length = 0 Then
                    Me.AddCriticalMessage("Service From Letter Prefix Missing")
                End If

                If Me.ValidatorMessages.CriticalMessages.Count > 0 Then
                    Throw New ValidationException(Me)
                End If
            End Sub

        End Class

    End Namespace

End Namespace